File: java/json-transform/src/test/java/co/nlighten/jsontransform/functions/TransformerFunctionAvgTest.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.BaseTest;
import com.google.gson.JsonNull;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

public class TransformerFunctionAvgTest extends BaseTest {
    @Test
    void inline() {
        var arr = new Object[] {4,2,13.45, JsonNull.INSTANCE};
        assertTransformation(arr, "$$avg():$", fromJson("4.8625"));
        assertTransformation(arr, "$$avg(1):$", fromJson("5.1125"));
    }

    static class Holder {
        final BigDecimal value;

        Holder(BigDecimal value) {
            this.value = value;
        }
    }
    @Test
    void object() {
        var arr = new Holder[] {new Holder(new BigDecimal(4)),new Holder( new BigDecimal(2)),new Holder( new BigDecimal("13.45")), new Holder(null)};


        assertTransformation(arr, fromJson("""
{
  "$$avg":"$","by":"##current.value"
}
"""), fromJson("4.8625"));

        assertTransformation(arr, fromJson("""
{
  "$$avg":"$","by":"##current.value","default":1
}
"""), fromJson("5.1125"));

    }

}

File: javascript/json-transform/src/__tests__/functions/TransformerFunctionAvg.test.ts
import { describe, test } from "vitest";
import { assertTransformation } from "../BaseTransformationTest";
import { BigDecimal } from "../../functions/common/FunctionHelpers";

class Holder {
  readonly value: number | null | bigint;

  constructor(value: number | null | bigint) {
    this.value = value;
  }
}

describe("TransformerFunctionAvg", () => {
  test("inline", async () => {
    const arr = [4, BigInt(2), 13.45, null];
    await assertTransformation(arr, "$$avg():$", BigDecimal(4.8625));
    await assertTransformation(arr, "$$avg(1):$", BigDecimal(5.1125));
  });

  test("object", async () => {
    const arr = [new Holder(BigInt(4)), new Holder(BigInt(2)), new Holder(13.45), new Holder(null)];
    await assertTransformation(arr, { $$avg: "$", by: "##current.value" }, BigDecimal(4.8625));
    await assertTransformation(arr, { $$avg: "$", by: "##current.value", default: 1 }, BigDecimal(5.1125));
  });
});
